### PR TITLE
EIP-3436 - Expanded Clique Block Choice Rule

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Implement the EIP-3436 expanded clique block choice rule.  Specifically, instead of taking the heaviest chain then first observed chain when presented with multiple chain heads the following choices are made in order:

1. Highest Total Difficulty
2. Shortest Chain
3. Most recent in-turn-proposer
4. Lowest Block hash

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java [hunk 1]
```diff
@@ -132,14 +134,18 @@ protected PluginServiceFactory createAdditionalPluginServices(final Blockchain b
...
+                epochManager,
+                blockInterface),
+            new VoteProposer(),
             epochManager,
-            blockInterface),
-        new VoteProposer(),
-        epochManager,
-        blockInterface);
+            blockInterface);
+    installCliqueBlockChoiceRule(blockchain, cliqueContext);
```
Comment (RatanRSur) [inline besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java:147]: Started going down a rabbit hole to remove this mutation. I didn't see a way out that isn't too intrusive.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java [hunk 2]
```diff
@@ -166,4 +167,15 @@ public boolean removeChainReorgObserver(final long observerId) {
...
+  @Override
+  public Comparator<BlockHeader> getBlockChoiceRule() {
+    return (a, b) -> {
+      throw new NonDeterministicOperationException(
+          "ReferenceTestBlockchian for VMTest Chains do not support fork choice rules");
+    };
+  }
+
+  @Override
+  public void setBlockChoiceRule(final Comparator<BlockHeader> blockChoiceRule) {}
```
Comment (RatanRSur) [inline ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java:180]: Could we add an exception here too? Or does it need to be a noop for things to work?

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java [hunk 3]
```diff
@@ -91,4 +95,38 @@ private static int minimumBlocksSincePreviousSigning(final ValidatorProvider val
...
+    blockchain.setBlockChoiceRule(
+        blockchain
+            // most total difficulty
+            .getBlockChoiceRule()
+            // shortest chain
+            .thenComparing(Comparator.comparing(ProcessableBlockHeader::getNumber).reversed())
+            // proposer with least recent in-turn block
+            .thenComparing(
+                Comparator.comparing(
+                    (BlockHeader header) -> distanceFromInTurn(header, cliqueContext)))
```
Comment (RatanRSur) [inline consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:110]:
```suggestion
            .thenComparing(header -> distanceFromInTurn(header, cliqueContext)))
```

**Review by RatanRSur — DISMISSED** [review]

**Comment by vmichalik** [comment]
@shemnon is this one on hold or do you need someone else to take a look at it?

**Comment by shemnon** [comment]
I was waiting on some ACD wide consensus on the issue, and the only interest I got from Geth was a comment where they mis-read the spec - https://ethereum-magicians.org/t/eip-3436-expanded-clique-block-choice-rule/5809/5

It's not a change one client should do in isolation, and ACD's agenda is fairly packed.  I expect it to fester until goerli halts again.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java (hunk 2 above)
Comment (shemnon) [reply ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestBlockchain.java]: Sure.  Assuming CI doesn't break.

File: besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java (hunk 1 above)
Comment (shemnon) [reply besu/src/main/java/org/hyperledger/besu/controller/CliqueBesuControllerBuilder.java:144]: The only way to be less intrusive is for a more invasive DI change, like going to Dagger.

**Comment by macfarla** [comment]
@shemnon is this one looking for a review or still waiting on ACD?

**Comment by shemnon** [comment]
Still needs that approval I asked for 3 months ago

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java [hunk 4]
```diff
@@ -90,4 +92,36 @@ private static int minimumBlocksSincePreviousSigning(
...
+
+  public static void installCliqueBlockChoiceRule(
+      final Blockchain blockchain, final CliqueContext cliqueContext) {
+    blockchain.setBlockChoiceRule(
+        blockchain
+            // most total difficulty
+            .getBlockChoiceRule()
+            // shortest chain
+            .thenComparing(Comparator.comparing(ProcessableBlockHeader::getNumber).reversed())
+            // proposer with least recent in-turn block
```
Comment (macfarla) [inline consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:104]: should this be "most recent" or is the PR description wrong?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/Blockchain.java [hunk 5]
```diff
@@ -230,4 +231,22 @@ default long observeLogs(final Consumer<LogWithMetadata> logObserver) {
    * @return {@code true} if the observer was removed; otherwise {@code false}
    */
   boolean removeChainReorgObserver(long observerId);
+
+  /**
+   * Gets the current block choice rule. When presented with two block headers indicate which chain
+   * is more preferred. greater than zero: the first chain is more preferred, less than zero: the
+   * second chain is more preferred, or zero: no preference.
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/Blockchain.java:238]: s/more preferred/preferred

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java [hunk 6]
```diff
@@ -90,4 +92,36 @@ private static int minimumBlocksSincePreviousSigning(
...
+            .thenComparing((BlockHeader header) -> distanceFromInTurn(header, cliqueContext))
+            // last resort: blockhash as uint256.
+            .thenComparing(Comparator.comparing(BlockHeader::getHash).reversed()));
+  }
+
+  static int distanceFromInTurn(final BlockHeader header, final CliqueContext context) {
```
Comment (macfarla) [inline consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:110]: reading the spec I'm still confused - maybe it's worth putting the formula in for context? 

Clients should choose the block with the largest value for this calculation:

(header_number - validator_index) % validator_count

**Review by macfarla — COMMENTED** [review]
couple of comments. I'd like someone from the mainnet team to review as well, it's a bit out of my area. sonarcloud is complaining but none of that was introduced by this PR

**Review by matkt — APPROVED** [review]
LGTM, just need to resolve @macfarla 's comments

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java (hunk 6 above)
Comment (shemnon) [reply consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:110]: added javadoc.

File: consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java (hunk 4 above)
Comment (shemnon) [reply consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueHelpers.java:104]: this flopped around several times.  current eip says maximum of the formula, so it's an inverted comparator.

**Pull request merged**